bluefog/common/operations.cc

```diff
@@ -281,6 +260,62 @@ Response ConstructResponse(MessageTable* message_table, std::string name) {
       break;
     }
   }
+  return error;
+}
+
+// Once a tensor is ready to be reduced, the coordinator sends an MPIResponse
+// instructing all ranks to start the reduction to all ranks. The MPIResponse
+// also contains error messages in case the submitted MPIRequests were not
+// valid (for example, contained mismatched shapes or types).
+//
+// Constructing the MPIResponse, thus, requires a whole lot of error checking.
+Response ConstructResponse(MessageTable* message_table, std::string name) {
+  bool error = false;
+  auto it = message_table->find(name);
+  assert(it != message_table->end());
+
+  std::vector<Request>& requests = std::get<0>(it->second);
+  assert(requests.size() > 0);
+  auto message_type = requests[0].request_type();
+
+  std::ostringstream error_message_stream;
+
+  // Make sure all requests' message type and data type are the same.
+  error = CheckRequestAndDataType(requests, error_message_stream);
+
+  // Make sure all requests use the same device.
+  if (!error) {
+    error = CheckRequestDevice(requests, error_message_stream);
+  }
+
+  // If we are doing a broadcast, check that all root ranks are identical.
+  if (!error) {
+    if (message_type == Request::BROADCAST) {
+      error = CheckRequestRootRank(requests, error_message_stream);
+    }
+  }
+
+  // If we are doing an (neighbor_)allreduce or broadcast, check that all tensor
+  // shapes are identical.
+  if (!error) {
+    if (message_type == Request::ALLREDUCE ||
+        message_type == Request::BROADCAST ||
+        message_type == Request::NEIGHBOR_ALLREDUCE ||
+        message_type == Request::WIN_CREATE) {
+      error = CheckRequestTensorShape(requests, error_message_stream);
+    }
+  }
+
+  // If we are doing (neighbor_)allgather, make sure all but the first dimension
+  // are the same. The first dimension may be different and the output tensor is
+  // the sum of the first dimension.
+  if (!error) {
+    if (message_type == Request::ALLGATHER ||
+        message_type == Request::NEIGHBOR_ALLGATHER) {
+      error = CheckRequestGatherTensorShape(requests, error_message_stream);
+    }
+  }
+
   std::vector<int32_t> devices(requests.size());
```

Comment:
Shall we skip the following if error occurs before?

Reply:
Right. Update.